File: unity/Assets/BarelyMusician/Scripts/Instrument.cs
using System;
using UnityEngine;

namespace Barely {
  /// A representation of a musical instrument that can be played in real-time.
  [RequireComponent(typeof(AudioSource))]
  public abstract class Instrument : MonoBehaviour {
    /// Audio source.
    public AudioSource Source { get; private set; } = null;

    /// Control event callback.
    ///
    /// @param index Control index.
    /// @param control Control value.
    public delegate void ControlEventCallback(int index, double value);
    public event ControlEventCallback OnControl;

    [Serializable]
    public class ControlEvent : UnityEngine.Events.UnityEvent<int, float> {}
    public ControlEvent OnControlEvent;

    /// Note control event callback.
    ///
    /// @param pitch Note pitch.
    /// @param index Note control index.
    /// @param control Note control value.
    public delegate void NoteControlEventCallback(double pitch, int index, double value);
    public event NoteControlEventCallback OnNoteControl;

    [Serializable]
    public class NoteControlEvent : UnityEngine.Events.UnityEvent<float, int, float> {}
    public NoteControlEvent OnNoteControlEvent;

    /// Note off event callback.
    ///
    /// @param pitch Note pitch.
    public delegate void NoteOffEventCallback(double pitch);
    public event NoteOffEventCallback OnNoteOff;

    [Serializable]
    public class NoteOffEvent : UnityEngine.Events.UnityEvent<float> {}
    public NoteOffEvent OnNoteOffEvent;

    /// Note on event callback.
    ///
    /// @param pitch Note pitch.
    /// @param intensity Note intensity.
    public delegate void NoteOnEventCallback(double pitch, double intensity);
    public event NoteOnEventCallback OnNoteOn;

    [Serializable]
    public class NoteOnEvent : UnityEngine.Events.UnityEvent<float, float> {}
    public NoteOnEvent OnNoteOnEvent;

    /// Returns a control value.
    ///
    /// @param index Control index.
    /// @return Control value.
    public double GetControl(int index) {
      return Musician.Internal.Instrument_GetControl(_id, index);
    }

    /// Returns a note control value.
    ///
    /// @param pitch Note pitch.
    /// @param index Control index.
    /// @return Control value.
    public double GetNoteControl(double pitch, int index) {
      return Musician.Internal.Instrument_GetNoteControl(_id, pitch, index);
    }

    /// Returns whether a note is on or not.
    ///
    /// @param Note pitch.
    /// @return True if on, false otherwise.
    public bool IsNoteOn(double pitch) {
      return Musician.Internal.Instrument_IsNoteOn(_id, pitch);
    }

    /// Resets all control values.
    public void ResetAllControls() {
      Musician.Internal.Instrument_ResetAllControls(_id);
    }

    /// Resets all note control values.
    ///
    /// @param pitch Note pitch.
    public void ResetAllNoteControls(double pitch) {
      Musician.Internal.Instrument_ResetAllNoteControls(_id, pitch);
    }

    /// Resets a control value.
    ///
    /// @param index Control index.
    public void ResetControl(int index) {
      Musician.Internal.Instrument_ResetControl(_id, index);
    }

    /// Resets a note control value.
    ///
    /// @param pitch Note pitch.
    /// @param index Note control index.
    public void ResetNoteControl(double pitch, int index) {
      Musician.Internal.Instrument_ResetNoteControl(_id, pitch, index);
    }

    /// Sets all notes off.
    public void SetAllNotesOff() {
      Musician.Internal.Instrument_SetAllNotesOff(_id);
    }

    /// Sets a control value.
    ///
    /// @param index Control index.
    /// @param value Control value.
    /// @param slopePerBeat Control slope in value change per beat.
    public void SetControl(int index, double value, double slopePerBeat = 0.0) {
      Musician.Internal.Instrument_SetControl(_id, index, value, slopePerBeat);
    }

    /// Sets data.
    ///
    /// @param dataPtr Pointer to data.
    /// @param size Data size in bytes.
    public void SetData(IntPtr dataPtr, int size) {
      Musician.Internal.Instrument_SetData(_id, dataPtr, size);
    }

    /// Sets a note control value.
    ///
    /// @param pitch Note pitch.
    /// @param index Note control index.
    /// @param value Note control value.
    /// @param slopePerBeat Note control slope in value change per beat.
    public void SetNoteControl(double pitch, int index, double value, double slopePerBeat = 0.0) {
      Musician.Internal.Instrument_SetNoteControl(_id, pitch, index, value, slopePerBeat);
    }

    /// Sets a note off.
    ///
    /// @param pitch Note pitch.
    public void SetNoteOff(double pitch) {
      Musician.Internal.Instrument_SetNoteOff(_id, pitch);
    }

    /// Sets a note on.
    ///
    /// @param pitch Note pitch.
    /// @param intensity Note intensity.
    public void SetNoteOn(double pitch, double intensity = 1.0) {
      Musician.Internal.Instrument_SetNoteOn(_id, pitch, intensity);
    }

    /// Class that wraps the internal api.
    public static class Internal {
      /// Internal control event callback.
      public static void OnControlEvent(Instrument instrument, int index, double value) {
        instrument.OnControl?.Invoke(index, value);
        instrument.OnControlEvent?.Invoke(index, (float)value);
      }

      /// Internal note control event callback.
      public static void OnNoteControlEvent(Instrument instrument, double pitch, int index,
                                            double value) {
        instrument.OnNoteControl?.Invoke(pitch, index, value);
        instrument.OnNoteControlEvent?.Invoke((float)pitch, index, (float)value);
      }

      public static void OnNoteOffEvent(Instrument instrument, double pitch) {
        instrument.OnNoteOff?.Invoke(pitch);
        instrument.OnNoteOffEvent?.Invoke((float)pitch);
      }

      /// Internal note on event callback.
      public static void OnNoteOnEvent(Instrument instrument, double pitch, double intensity) {
        instrument.OnNoteOn?.Invoke(pitch, intensity);
        instrument.OnNoteOnEvent?.Invoke((float)pitch, (float)intensity);
      }
    }

    protected virtual void Awake() {
      Source = GetComponent<AudioSource>();
      Source.clip = AudioClip.Create("Ones", 64, 1, AudioSettings.outputSampleRate, false);
      float[] ones = new float[64];
      for (int i = 0; i < ones.Length; ++i) {
        ones[i] = 1.0f;
      }
      Source.clip.SetData(ones, 0);
      Source.loop = true;
    }

    protected virtual void OnDestroy() {
      Source = null;
    }

    protected virtual void OnEnable() {
      Musician.Internal.Instrument_Create(this, ref _id);
      Source?.Play();
    }

    protected virtual void OnDisable() {
      Source?.Stop();
      Musician.Internal.Instrument_Destroy(ref _id);
    }

    private void OnAudioFilterRead(float[] data, int channels) {
      Musician.Internal.Instrument_Process(_id, data, channels);
    }

    // Identifier.
    private Int64 _id = Musician.Internal.InvalidId;
  }
}

File: platforms/wasm/src/instrument.js
import {CommandType} from './context.js'

/**
 * A representation of a musical instrument that can be played in real-time.
 */
export class Instrument {
  /**
   * @param {!Engine} engine
   * @param {number} handle
   * @param {function(number):void} noteOnCallback
   * @param {function(number):void} noteOffCallback
   */
  constructor(engine, handle, noteOnCallback, noteOffCallback) {
    /** @private @const {!Engine} */
    this._engine = engine;

    /** @private @const {number} */
    this._handle = handle;

    /** @public */
    this.noteOnCallback = noteOnCallback;

    /** @public */
    this.noteOffCallback = noteOffCallback;
  }

  /**
   * Destroys the instrument.
   */
  destroy() {
    this._engine._instruments.delete(this._handle);
    this._engine._pushCommand({type: CommandType.INSTRUMENT_DESTROY, handle: this._handle});
  }

  /**
   * Sets all notes off.
   */
  setAllNotesOff() {
    this._engine._pushCommand(
        {type: CommandType.INSTRUMENT_SET_ALL_NOTES_OFF, handle: this._handle});
  }

  /**
   * Sets a control value.
   * @param {number} typeIndex
   * @param {number} value
   */
  setControl(typeIndex, value) {
    this._engine._pushCommand(
        {type: CommandType.INSTRUMENT_SET_CONTROL, handle: this._handle, typeIndex, value});
  }

  /**
   * Sets a note control value.
   * @param {number} pitch
   * @param {number} typeIndex
   * @param {number} value
   */
  setNoteControl(pitch, typeIndex, value) {
    this._engine._pushCommand({
      type: CommandType.INSTRUMENT_SET_NOTE_CONTROL,
      handle: this._handle,
      pitch,
      typeIndex,
      value
    });
  }

  /**
   * Sets a note off.
   * @param {number} pitch
   */
  setNoteOff(pitch) {
    this._engine._pushCommand(
        {type: CommandType.INSTRUMENT_SET_NOTE_OFF, handle: this._handle, pitch});
  }

  /**
   * Sets a note on.
   * @param {number} pitch
   * @param {number=} gain
   * @param {number=} pitchShift
   */
  setNoteOn(pitch, gain = 1.0, pitchShift = 0.0) {
    this._engine._pushCommand(
        {type: CommandType.INSTRUMENT_SET_NOTE_ON, handle: this._handle, pitch, gain, pitchShift});
  }

  /**
   * Loads and assigns sample data to the instrument.
   * @param {!Array<{pitch: number, url: string}>} sampleData
   * @return {!Promise<void>}
   */
  async setSampleData(sampleData) {
    const slices = [];

    for (const {pitch, url} of sampleData) {
      const response = await fetch(url);
      if (!response.ok) {
        console.warn(`Invalid sample data for pitch ${pitch}`);
        continue;
      }

      const arrayBuffer = await response.arrayBuffer();
      const audioBuffer = await this._engine._audioContext.decodeAudioData(arrayBuffer);

      slices.push({
        rootPitch: pitch,
        sampleRate: audioBuffer.sampleRate,
        samples: audioBuffer.getChannelData(0),  // mono only
      });
    }

    this._engine._pushCommand(
        {type: CommandType.INSTRUMENT_SET_SAMPLE_DATA, handle: this._handle, slices});
  }

  /** @return {number} */
  get handle() {
    return this._handle;
  }
}
